### PR TITLE
Add StreamStore for thread-safe storage of states in a params object

### PR DESCRIPTION
**Pull request by sethrj**
Until a more robust solution is implemented in #392, this is a solution for managing a vector of states (one per stream ID) and allocating them on the fly.

### CODE REVIEW

**Comment by sethrj** [comment]
The last commit is to fix a weird error when building CUDA that includes the collection mirror:
```
/var/jenkins/workspace/Celeritas_PR-722/src/corecel/data/CollectionMirror.hh(131): error #128-D: loop is not reachable
          detected during:
            instantiation of "const P<celeritas::Ownership::const_reference, M> &celeritas::CollectionMirror<P>::ref<M>() const [with P=celeritas::ParticleParamsData, M=celeritas::MemSpace::host]" 
(77): here
            instantiation of "const celeritas::CollectionMirror<P>::HostRef &celeritas::CollectionMirror<P>::host_ref() const [with P=celeritas::ParticleParamsData]" 
(75): here
            instantiation of "const celeritas::CollectionMirror<P>::HostRef &celeritas::CollectionMirror<P>::host() const [with P=celeritas::ParticleParamsData]" 
/var/jenkins/workspace/Celeritas_PR-722/src/celeritas/phys/ParticleParams.hh(97): here
```
and CollectionMirror.hh(131) is the `CELER_ASSERT_UNREACHABLE` macro.

File: src/corecel/data/StreamStore.hh [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+#include "CollectionStateStore.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Helper class for storing parameters and multiple stream-dependent states.
+ *
+ * This requires a templated ParamsData and StateData. Hopefully this
+ * frankenstein of a class will be replaced by a
```
Comment (amandalund) [inline src/corecel/data/StreamStore.hh]: ...by a what? ;)

**Review by amandalund — APPROVED** [review]
Thanks @sethrj, seems like a good solution for now!

File: src/corecel/data/StreamStore.hh (hunk 1 above)
Comment (sethrj) [reply src/corecel/data/StreamStore.hh]: by a squirrel!

Comment (sethrj) [reply src/corecel/data/StreamStore.hh]: ![https://media.tenor.com/UJYClb5wTboAAAAC/squirrel-doug.gif]

Comment (sethrj) [reply src/corecel/data/StreamStore.hh]: Squirrel!

**Pull request merged**